pkg/daemon/controller/controller.go

```diff
@@ -64,58 +73,132 @@ func RunController(kubeconfigPath string) error {
 		return errors.Wrap(err, "error using kubeconfig to build config")
 	}
 
-	// TODO: This is a client that reads directly from the server, not a cached client like used in WMCO. This should
-	//       be replaced with the cache client received by mgr.GetClient()
 	clientScheme := runtime.NewScheme()
 	err = clientgoscheme.AddToScheme(clientScheme)
 	if err != nil {
 		return err
 	}
-	clientset, err := client.New(config, client.Options{Scheme: clientScheme})
+	// This is a client that reads directly from the server, not a cached client. This is required to be used here, as
+	// the cached client, created by ctrl.NewManager() will not be functional until the manager is started.
+	directClient, err := client.New(config, client.Options{Scheme: clientScheme})
 	if err != nil {
 		return err
 	}
 
 	// use the client to find the name of the node associated with the VM this is running on
+	ctx := ctrl.SetupSignalHandler()
 	var nodes core.NodeList
-	err = clientset.List(context.TODO(), &nodes)
+	err = directClient.List(ctx, &nodes)
 	if err != nil {
 		return err
 	}
 	addrs, err := localInterfaceAddresses()
 	if err != nil {
 		return err
 	}
-	node, err := currentNode(&nodes, addrs)
+	node, err := findNodeByAddress(&nodes, addrs)
+	if err != nil {
+		return err
+	}
+
+	ctrlMgr, err := ctrl.NewManager(config, ctrl.Options{
+		Namespace: "openshift-windows-machine-config-operator",
+		Scheme:    clientScheme,
+	})
 	if err != nil {
+		return errors.Wrap(err, "unable to start manager")
+	}
+	sc := NewServiceController(ctx, ctrlMgr.GetClient(), svcMgr, node.Name)
+	if err = sc.SetupWithManager(ctrlMgr); err != nil {
 		return err
 	}
+	klog.Info("Starting manager, awaiting events")
+	if err := ctrlMgr.Start(ctx); err != nil {
+		return err
+	}
+	return nil
+}
+
+// NewServiceController returns a pointer to a ServiceController object
+func NewServiceController(ctx context.Context, client client.Client, mgr winsvc.Mgr, nodeName string) *ServiceController {
+	return &ServiceController{client: client, Mgr: mgr, ctx: ctx, nodeName: nodeName,
+		watchNamespace: "openshift-windows-machine-config-operator"}
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (sc *ServiceController) SetupWithManager(mgr ctrl.Manager) error {
+	nodePredicate := predicate.Funcs{
+		// A node's name will never change, so it is fine to use the name for node identification
+		CreateFunc: func(e event.CreateEvent) bool {
+			return sc.nodeName == e.Object.GetName()
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			// Only process update events if the desired version has changed
+			return sc.nodeName == e.ObjectNew.GetName() &&
+				e.ObjectOld.GetAnnotations()[desiredVersionAnnotation] != e.ObjectNew.GetAnnotations()[desiredVersionAnnotation]
+		},
+		GenericFunc: func(e event.GenericEvent) bool {
+			return sc.nodeName == e.Object.GetName()
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
```

Comment:
would the delete come into play with the cleanup command?

Reply:
I'm not sure at this moment, for now I'll change it to drop delete events as they dont have a purpose at the moment.